topics/07_web_ui_testing/main_web_automation_with_playwright.md

```diff
@@ -1,3 +1,383 @@
+
+#### 'expect' vs default Python 'assert'
+- 'assert' returns Boolean
+- 'expect' returns None
+
+example:
+```python
+    def is_macros_not_displayed_on_page(self, macro_name: str) -> None:
+        expect(self.page.locator(f"//*[text()[contains(.,'{macros_name}')]]")).not_to_be_visible()
+```
+compare with:
+```python
+    def is_macros_not_displayed_on_page(self, macro_name: str) -> bool:
+        return self.page.locator(f"//*[text()[contains(.,'{macros_name}')]]").not_to_be_visible()
+```
+The last one can be used in test in way like:
+
+```python
+assert request_macros_page.is_macro_displayed_on_page(setup_ui_data["macro_name"])
+```
+It more convenient for code but sometimes not convenient in case of delay with page loading
+
+
+
+### Additional features of Playwright vs Pytest
+
+1. Playwright supports all standart types of locators like css or xpath. Additionally it provides the additional functionality like last and first locator:
+
+```python
+page.locator(".classname").last.click()
+```
+2. Ability operate with locator in "#shadow-root"
+Remark: #shadow-root is area within DOM. Locators within this section are not available for many Test Automation Frameworks
+
+3. Playwright has numerous build-in methods like:
+3.1
+   .to_have_text()
+```python
+page.get_by_role("#Category")).to_have_text(["plates", "caps", "catleriese"])
+```
+3.2
+   .get_by_text()
+```python
+page.get_by_text("orange").click()
+```
+
+### Action methods
+
+```python
+    # Actions on page
+    def navigate(self, url: str):
+        self.page.goto(url)
+
+
+    # Get information from page
+    def check_loading_initial_page(self):
+        return self.text_selenium.is_visible()
+```
+
+### Example of code
+
+#### Page object
+staring_page.py
+```python
+from playwright.sync_api import Page
+
+class StartingPage:
+    # Initialize page and selectors
+    def __init__(self, page: Page):
+        self.page = page
+        self.text_selenium = page.locator("[alt='Selenium Online Training']")
+
+
+    # Actions on page
+    def navigate(self, url: str):
+        self.page.goto(url)
+
+
+    # Get information from page
+    def check_loading_initial_page(self):
+        return self.text_selenium.is_visible()
+```
+
+#### Test suit
+test_starting_page.py
+```python
+import pytest
+from starting_page import StartingPage
+from playwright.sync_api import sync_playwright
+
+# Define a pytest fixture to set up Playwright and browser instance
+@pytest.fixture(scope="session")
+def browser():
+    with sync_playwright() as playwright:
+        browser = playwright.chromium.launch(headless=False)
+        yield browser
+        browser.close()
+
+
+@pytest.fixture
+def page(browser):
+    context = browser.new_context()
+    page = context.new_page()
+    return page
+
+
+# Test case using the StartingPage object
+def test_login(page):
+    login_page = StartingPage(page)
+    login_page.navigate("https://demoqa.com")
+```
+
+## Hometask
```

Comment:
We are using `Task Exercises:` for the homework blocks. See the previous topics.